### PR TITLE
Storage: Populate custom volume snapshot creation date

### PR DESCRIPTION
**Pull request by roosterfish**
When generating a custom volume's backup config, make sure it's creation date is included in the struct. This is currently causing snapshots of copied volumes to not have any creation date:

```bash
julian@thinkpad:~/dev/lxd/test$ lxc storage volume show dir vol2/snap0
description: ""
expires_at: 0001-01-01T00:00:00Z
name: snap0
config: {}
content_type: filesystem
created_at: 0001-01-01T00:00:00Z
```

Additionally this change ensures that when refreshing custom storage volumes and their snapshots, LXD does not re-transfer all of the snapshots. This was caused by:
* the migration header protocol not sending the snapshots creation date
* and the comparison on the target side using different granularity for both source and target timestamps

This PR adds a check for both custom storage volume and instance snapshots. But the date is only missing for custom volume snapshots. It was observed in https://github.com/canonical/lxd/pull/12840.

The fact that LXD refreshes all the snapshots when migration custom storage volumes was first discovered here https://github.com/canonical/lxd/issues/12668.

### CODE REVIEW

**Comment by roosterfish** [comment]
The commit https://github.com/canonical/lxd/pull/12893/commits/d94fdac5e12abfb7e6ca929de125ede0a6e02efa is required since the test is checking the behavior of the bug.

Previously both the snapshots of `vol5` and `vol6` did have the creation timestamp `0001-01-01T00:00:00Z`. Now when doing `lxc storage volume cp vol6 vol5 --refresh` LXD will not replace the snapshots on `vol5` since they have both the same (wrong) creation timestamp and name.

With the changes of this PR, the snapshots of `vol5` have an earlier creation time than the ones of `vol6`. This will effectively cause a refresh of the snapshots of `vol5`. This is the expected behavior and reason why the test requires a change.

**Comment by roosterfish** [comment]
@tomponline ready for another look :)

File: lxd/migrate_storage_volumes.go [hunk 1]
```diff
@@ -365,8 +365,13 @@ func (c *migrationSink) DoStorage(state *state.State, projectName string, poolNa
...
 			targetSnapshotsComparable = append(targetSnapshotsComparable, storagePools.ComparableSnapshot{
-				Name:         targetSnapName,
-				CreationDate: targetSnap.CreationDate,
+				Name: targetSnapName,
+
+				// The list of source snapshots from the offer header
+				// contains the creation timestamps in seconds granularity.
+				// Also use second based granularity for the target snapshots to be able to compare them.
+				// They are stored with nanoseconds in the database.
+				CreationDate: targetSnap.CreationDate.Round(time.Second),
```
Comment (tomponline) [inline lxd/migrate_storage_volumes.go:374]: Hrm, do you think we may get problems with incorrect comparisons (triggering unnecessary refreshes) if the time.Unix() conversion used on the sending side does a different rounding (or non at all) to that which time.Round() uses here?

Perhaps we should convert to unix timestamp using time.Unix() and then back into a timestamp?

**Review by tomponline — COMMENTED** [review]
Apart from the question about rounding, rest looks good thanks

**Comment by roosterfish** [comment]
> Apart from the question about rounding, rest looks good thanks

The change with the unix timestamp is in. Additionally I have removed sending the snapshots expiry date in the migration header as this will value won't be used at all and instead picked from the marshalled JSON that gets sent afterwards.

**Review by tomponline — APPROVED** [review]

**Pull request merged**